pkg/blueprint/firstboot_customizations.go

```diff
@@ -0,0 +1,154 @@
+	CACerts []string `json:"cacerts,omitempty" toml:"cacerts,omitempty"`
+
+	// Registration command as generated by the Satellite server. Required, if
+	// type is set to "satellite".
+	Command string `json:"command,omitempty" toml:"command,omitempty"`
+}
+
+// AAPFirstbootCustomization contains fields specific to AAP firstboot
+// customizations. All fields must be prefixed with "AAP".
+type AAPFirstbootCustomization struct {
+	FirstbootCommonCustomization
+
+	// Optional CA certificate to enroll into the system before executing the
+	// firstboot script.
+	CACerts []string `json:"cacerts,omitempty" toml:"cacerts,omitempty"`
+
+	// Job template URL as generated by the AAP server. Required if type is set
+	// to "aap". Example URLs are
+	// https://aap.example.com/api/controller/v2/job_templates/9/callback/ or
+	// https://aap.example.com/api/v2/job_templates/9/callback/ depending on the
+	// AAP version.
+	JobTemplateURL string `json:"job_template_url,omitempty" toml:"job_template_url,omitempty"`
+
+	// The host config key. Required if type is set to "aap".
+	HostConfigKey string `json:"host_config_key,omitempty" toml:"host_config_key,omitempty"`
+}
+
+func (t FirstbootScriptCustomization) MarshalJSON() ([]byte, error) {
+	b, err := t.union.MarshalJSON()
+	return b, err
+}
+
+func (t *FirstbootScriptCustomization) UnmarshalJSON(b []byte) error {
+	err := t.union.UnmarshalJSON(b)
+	return err
+}
+
+func (t FirstbootScriptCustomization) MarshalTOML() ([]byte, error) {
+	b, err := t.union.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	return jsonToToml(b)
+}
+
+func (t *FirstbootScriptCustomization) UnmarshalTOML(data any) error {
+	return unmarshalTOMLviaJSON(t, data)
+}
+
+var ErrUnknownFirstbootCustomization = errors.New("unknown firstboot customization")
+
+// SelectUnion returns the specific firstboot customization types. The detection
+// is based on the "type" field in the JSON payload. Returns nil for irrelevant types,
+// or if the type is unknown or if there is an error during unmarshaling.
+func (sp FirstbootScriptCustomization) SelectUnion() (*CustomFirstbootCustomization, *SatelliteFirstbootCustomization, *AAPFirstbootCustomization, error) {
+	var fcc FirstbootCommonCustomization
+	err := json.Unmarshal(sp.union, &fcc)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	switch strings.ToLower(fcc.Type) {
+	case "custom":
+		var cc CustomFirstbootCustomization
+		err = json.Unmarshal(sp.union, &cc)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		return &cc, nil, nil, nil
+	case "satellite":
+		var sc SatelliteFirstbootCustomization
+		err = json.Unmarshal(sp.union, &sc)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		return nil, &sc, nil, err
+	case "aap":
+		var ac AAPFirstbootCustomization
+		err = json.Unmarshal(sp.union, &ac)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		return nil, nil, &ac, err
+	default:
+		return nil, nil, nil, ErrUnknownFirstbootCustomization
+	}
+
+}
```

Comment:
One thing this doesn't catch is cross-type fields being set.  For example, if `type == "custom"` and `job_template_url != ""`, that should be invalid.  We can catch this automatically with strict unmarshalling (`DisallowUnknownFields`) and I think that would go a long way to avoiding silently ignoring invalid configs which can lead to confusion.